nvflare/app_common/job_schedulers/job_scheduler.py

```diff
@@ -24,99 +24,135 @@
-            resource_reqs=job.resource_spec, resource_check_results=resource_check_results, fl_ctx=fl_ctx
-        )
-
-    required_sites_received = 0
-    num_sites_ok = 0
-    sites_dispatch_info = {}
-    for site_name, check_result in resource_check_results.items():
-        if check_result[0]:
-            sites_dispatch_info[site_name] = DispatchInfo(
-                resource_requirements=job.resource_spec[site_name], token=check_result[1]
+    def _try_job(self, job: Job, fl_ctx) -> (bool, Optional[Dict[str, DispatchInfo]]):
+        engine = fl_ctx.get_engine()
+        online_clients = engine.get_clients()
+        online_site_names = [x.name for x in online_clients]
+
+        if not job.deploy_map:
+            raise RuntimeError(f"Job ({job.job_id}) does not have deploy_map, can't be scheduled.")
+
+        # deploy_map: {"app_name": []} will be treated as deploying to all online clients
+        all_deploy_map_keys = list(job.deploy_map.keys())
+        if len(all_deploy_map_keys) == 1 and len(job.deploy_map[all_deploy_map_keys[0]]) == 0:
+            applicable_sites = online_site_names
+            sites_to_app = {x: all_deploy_map_keys[0] for x in applicable_sites}
+        else:
+            applicable_sites = []
+            sites_to_app = {}
+            for app_name in job.deploy_map:
+                for site_name in job.deploy_map[app_name]:
+                    if site_name in online_site_names:
+                        applicable_sites.append(site_name)
+                        sites_to_app[site_name] = app_name
+        self.log_debug(fl_ctx, f"Job {job.job_id} is checking against applicable sites: {applicable_sites}")
+
+        if job.required_sites:
+            for s in job.required_sites:
+                if s not in applicable_sites:
+                    self.log_debug(fl_ctx, f"Job {job.job_id} can't be scheduled: required site {s} is not connected.")
+                    return False, None
+
+        if job.min_sites and len(applicable_sites) < job.min_sites:
+            self.log_debug(
+                fl_ctx,
+                f"Job {job.job_id} can't be scheduled: connected sites ({len(applicable_sites)}) "
+                f"are less than min_sites ({job.min_sites}).",
+            )
+            return False, None
+
+        # we are assuming server resource is sufficient
+        resource_reqs = {}
+        for site_name in applicable_sites:
+            if site_name in job.resource_spec:
+                resource_reqs[site_name] = job.resource_spec[site_name]
+            else:
+                resource_reqs[site_name] = {}
+        resource_check_results = self._check_client_resources(resource_reqs=resource_reqs, fl_ctx=fl_ctx)
+
+        if not resource_check_results:
+            self.log_debug(fl_ctx, f"Job {job.job_id} can't be scheduled: resource check results is None or empty.")
+            return False, None
+
+        required_sites_not_enough_resource = list(job.required_sites)
```

Comment:
These lines: https://github.com/NVIDIA/NVFlare/blob/dev-2.1/nvflare/apis/job_def.py#L81-L82 should prevent that.

But I've added some checks here in case the implementation there changed.